20210530-acme-zhiwei.md

```diff
@@ -0,0 +1,200 @@
+
+Another plausible alternative to storing certificates is using TXT records. Instead of storing tls certificates as CERT RR, they can be stored as TXT Records under a well-known domain (`dns.coredns.io`) and can be used for serving DoH.
+
+### Clustered Environment
+
+This could be put into another separate plugin like `acme-worker` and `acme-leader`. However, I propose for it to stay under the `acme` plugin
+because it would share the code and logic needed for `acme`.
+
+In clustered environment, where there are multiple CoreDNS instances running, certificate sharing is a major concern. Certificate sharing is important as it is not feasible for each CoreDNS instance to fetch certificate individually.
+
+Since high availability is not a concern, a solution is to configure CoreDNS instances as "leader" and "worker". 
+
+1. Leader: In a cluster, a single DNS Server can be configured as "leader", which would be responsible for actually performing ACME challenges and obtaining certificate from the Let's Encrypt CA. The leader would then store the certificate in the zone file as CERT RR under the domain of `dns.coredns.io`. This certificate can be used to serve DoH whenever there is a DNS query to the master.
+
+2. Worker: The worker is responsible for fetching the certificates from the master. The worker is not responsible for performing ACME challenges. Whenever a DNS query is received at the worker node, the worker first queries the master for the certificate. Workers should cache the certificate and use it for further DoH queries.
+
+Configurations:
+
+Leader
+```
+https://example.com {
+  acme <DOMAIN> {
+    challenge <HTTP/DNS Challenge>
+    leader
+  }
+}
+```
+
+Worker
+```
+https://example.com {
+  acme <DOMAIN> {
+    worker
+  }
+}
+```
+
+The worker caches the CERT RR (for X seconds) to reduce querying the master for the certificate every time a DNS query is received.
+
+#### Cert Renewal in Clustered Environment
+
+In order to prevent the workers from using the stale certificates, the master disseminates the renewed certificate to them instead of workers polling the master for the certificates.
+
+## Timeline
+Due to the tight timeline, 50% of the project has to be implemented by July 11th.
+Period: June 1st - August 31st
+
+**31 May - 6 June**
+1. Approval of RFC
+2. Setting up of development environment
+3. First Meeting
+4. After approval of RFC, start of writing
```

Comment:
I would suggest the plan to be "start implementation now, while at the same time, updating and polishing RFC until it is merged". Implementation does not has to be done after RFC is merged.

Reply:
Already started implementing.